open-craft/opencraft
# Add support for multiple databases.

This PR adds support for defining multiple MySQL and MongoDB servers in the OC IM database. When creating a new instance, a single MySQL and MongoDB database server is assigned to it. If the instance is configured to use persistent databases, it will create necessary databases on its MySQL and MongoDB servers as part of the provisioning process for a new app server. If no database servers are present in the database when creating an instance, default MySQL and MongoDB servers will automatically be created using information from the `DEFAULT_INSTANCE_MYSQL_URL` and `DEFAULT_INSTANCE_MONGO_URL` environment variables.

cf. [OC-1966](https://tasks.opencraft.com/browse/OC-1966)

**Test instructions**

1. Checkout this branch and run new migrations via `make migrate`.

2. Set `DEFAULT_INSTANCE_MYSQL_URL` and `DEFAULT_INSTANCE_MONGO_URL` in your `.env` file. These settings replace `INSTANCE_MYSQL_URL` and `INSTANCE_MONGO_URL`, respectively; use values from OC IM stage `.env` for testing.

3. As a precondition check, verify that there are no `MySQLServer` or `MongoDBServer` objects in the database. You can do this by going to the Django admin (http://localhost:5000/admin/instance/mysqlserver/, http://localhost:5000/admin/instance/mongodbserver/) or via the shell:

    ```python
    assert MySQLServer.objects.count() == 0
    assert MongoDBServer.objects.count() == 0
    ```

4. Create a new instance that uses persistent databases:

    ```python
    from instance.factories import production_instance_factory
    production_instance = production_instance_factory(name='Production instance', sub_domain='production-instance-IiZei2se')
    ```

5. Verify that the database contains a single `MySQLServer` and a single `MongoDBServer`, and that the fields of these servers match the info from `DEFAULT_INSTANCE_MYSQL_URL` and `DEFAULT_INSTANCE_MONGO_URL`. You can do this by going to the Django admin or via the shell:

    ```python
    assert MySQLServer.objects.count() == 1
    assert MongoDBServer.objects.count() == 1
    
    from django.conf import settings
    from urllib.parse import urlparse

    mysql_url_obj = urlparse(settings.DEFAULT_INSTANCE_MYSQL_URL)
    mongo_url_obj = urlparse(settings.DEFAULT_INSTANCE_MONGO_URL)

    mysql_server = MySQLServer.objects.get()
    mongodb_server = MongoDBServer.objects.get()

    assert mysql_server.hostname == mysql_url_obj.hostname
    assert mysql_server.port == 3306
    assert mysql_server.username == mysql_url_obj.username
    assert mysql_server.password == mysql_url_obj.password

    assert mongodb_server.hostname == mongo_url_obj.hostname
    assert mongodb_server.port == 27017
    assert mongodb_server.username == mongo_url_obj.username
    assert mongodb_server.password == mongo_url_obj.password
    ```

6. Verify that the instance is configured to use the newly created database servers:

    ```python
    assert production_instance.mysql_server == mysql_server
    assert production_instance.mongodb_server == mongodb_server
    ```

7. Spawn a new app server for the instance:

    ```python
    production_instance.spawn_appserver()
    ```

    Watch the output for a few seconds; it should contain (and not error out at) the following lines:

    ```
    instance.models.instance  | instance=90 (Production inst) | Provisioning MySQL database...
    instance.models.instance  | instance=90 (Production inst) | Provisioning MongoDB databases...
    ```

    If you get Swift-related errors, just ignore them: Using Swift containers requires additional setup, and this PR does not touch Swift-related functionality/code, so it's not necessary to get Swift to work locally for the purpose of testing this PR.



8. To further test database server selection, create a few additional database servers and instances, then verify that each of the newly created instances is associated with one of the existing `MySQLServer` and `MongoDBServer` objects:

    ```python
    for i in range(3):
        MySQLServer.objects.create(
            hostname='mysql-server-{}'.format(i),
            username='mysql-user-{}'.format(i),
            password='mysql-pass-{}'.format(i),
            port=i,
        )
        MongoDBServer.objects.create(
            hostname='mongodb-server-{}'.format(i),
            username='mongodb-user-{}'.format(i),
            password='mongodb-pass-{}'.format(i),
            port=i,
        )

    for i in range(3):
        production_instance = production_instance_factory(
            name='Production instance {}'.format(i), 
            sub_domain='production-instance-{}'.format(i),
        )
        assert production_instance.mysql_server in MySQLServer.objects.all()
        assert production_instance.mongodb_server in MongoDBServer.objects.all()
    ```



**Reviewers**
